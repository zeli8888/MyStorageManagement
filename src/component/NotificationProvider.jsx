import { SnackbarProvider } from 'notistack';
import { useSnackbar } from 'notistack';
import notificationService from '../service/NotificationService';
import React, { useEffect } from 'react';
function InitializeNotificationService() {
    const { enqueueSnackbar } = useSnackbar();
    useEffect(() => {
        notificationService.initialize({ enqueueSnackbar });

        return () => {
            notificationService.reset();
        };
    }, [enqueueSnackbar]);

    return null;
}

const NotificationProvider = ({ children }) => {
    return <SnackbarProvider
        maxSnack={3}
        autoHideDuration={3000}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'left' }}
    >
        <InitializeNotificationService />
        {children}
    </SnackbarProvider>;
}

export default NotificationProvider