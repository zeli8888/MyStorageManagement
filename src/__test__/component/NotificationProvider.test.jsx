import { describe, expect, test, vi, beforeEach, afterEach } from 'vitest'
import { render, screen, cleanup } from '@testing-library/react'
import NotificationProvider from '../../component/NotificationProvider'
import notificationService from '../../service/NotificationService'
import { SnackbarProvider } from 'notistack'
import '@testing-library/jest-dom/vitest'

// Mock notification service module
vi.mock('../../service/NotificationService', () => ({
    default: {
        initialize: vi.fn(),
        reset: vi.fn()
    }
}))

// Mock notistack SnackbarProvider component for props verification
vi.mock('notistack', () => ({
    SnackbarProvider: vi.fn(({ children }) => <div>{children}</div>),
    useSnackbar: vi.fn(() => ({
        enqueueSnackbar: vi.fn()
    }))
}))

// Mock child component for rendering verification
const MockChild = () => <div>Test Child</div>

describe('NotificationProvider Component Tests', () => {
    beforeEach(() => {
        // Reset mock state before each test
        vi.clearAllMocks()
    })

    afterEach(cleanup)

    test('should render SnackbarProvider with correct props', () => {
        const { asFragment } = render(
            <NotificationProvider>
                <MockChild />
            </NotificationProvider>
        )
        expect(asFragment()).toMatchSnapshot()

        // Verify SnackbarProvider receives correct props
        expect(SnackbarProvider.mock.calls[0][0]).toEqual(
            expect.objectContaining({
                maxSnack: 3,
                autoHideDuration: 3000,
                anchorOrigin: { vertical: 'bottom', horizontal: 'left' }
            })
        )

        // Verify child component rendering
        expect(screen.getByText('Test Child')).toBeInTheDocument()
    })

    test('should initialize notification service on mount', () => {
        const { unmount } = render(<NotificationProvider><MockChild /></NotificationProvider>)

        // Verify initialize method is called
        expect(notificationService.initialize).toHaveBeenCalledWith({
            enqueueSnackbar: expect.any(Function)
        })

        // Simulate component unmount
        unmount()

        // Verify reset method is called on unmount
        expect(notificationService.reset).toHaveBeenCalled()
    })

    test('should reset notification service on component unmount', () => {
        const { unmount } = render(<NotificationProvider><MockChild /></NotificationProvider>)

        // Ensure initial state
        expect(notificationService.reset).not.toHaveBeenCalled()

        // Trigger unmount
        unmount()

        // Verify reset method is called
        expect(notificationService.reset).toHaveBeenCalledTimes(1)
    })
})
