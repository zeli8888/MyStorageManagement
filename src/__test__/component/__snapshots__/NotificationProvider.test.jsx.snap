// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationProvider Component Tests > should render SnackbarProvider with correct props 1`] = `
<DocumentFragment>
  <div>
    <div>
      Test Child
    </div>
  </div>
</DocumentFragment>
`;
